### ISSUE TITLE
Correct error reporting in _twistd_unix.py

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8731]
|[<img alt="markrwilliams's avatar" src="https://avatars.githubusercontent.com/u/704247?s=50" width="50" height="50">](https://github.com/markrwilliams)| @markrwilliams reported|
|-|-|
|Trac ID|trac#8731|
|Type|defect|
|Created|2016-08-07 19:42:44Z|
|Branch|https://github.com/twisted/twisted/tree/8731-bytes-repr-with-b|

(was "Explicit bytes repr utility for Python 2 and 3")

`twistd` daemonizes under Python 3 now that evgeni closed [#8155](https://github.com/twisted/twisted/issues/8155).  Thanks, evgeni!

However, if an exception occurs in the child and it attempts to write that exception to its status pipe, `twistd` will fail on Python 3, because the exception's message will be a `str` and not bytes.  It will also fail on Python 2 if the exception message is unicode.

The message as extracted from the exception also does not include the exception's name.  If the code uses `traceback.format_exception_only`, then the child can write the exception out with the same formatting it would have if it were the final line in a traceback.  On Python 2, `format_exception_only` happens to encode `unicode` messages with `ascii`/`backslashescape`.  Using it would solve the encoding problem and make the message clearer.

Finally, the parent/daemonization watcher should print any status message read off the pipe as a bytes repr.  This obviates the need to decode the message and will make more sense when it's encoded with `backslashescape`d.

As Adi explains [here](https://github.com/twisted/twisted/pull/461#issuecomment-238083180), `twisted.python.compat` should gain a utility function that always produces reprs that begin with `b` for byte strings regardless of the Python version:

`future_bytes_repr(b'a')` -> `b'a'`

`_twistd_unix` can then use this to format the bytes repr.

<details><summary>Searchable metadata</summary>

```
trac-id__8731 8731
type__defect defect
reporter__markrwilliams markrwilliams
priority__normal normal
milestone__Python_3_x Python-3.x
branch__8731_bytes_repr_with_b 8731-bytes-repr-with-b
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__None None
time__1470598964287651 1470598964287651
changetime__1470715047476403 1470715047476403
version__None None
owner__markrwilliams markrwilliams

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_1"></a>|@adiroiban commented|
|-|-|

and as part of this ticket we should also put this new API in use by fixing the twistd non-ascii stderr handling...

I think that this ticket should focus on fixing that bug and create the compat helper as part of the fix

**Comment by twisted-trac** [comment]
|[<img alt="markrwilliams's avatar" src="https://avatars.githubusercontent.com/u/704247?s=50" width="50" height="50">](https://github.com/markrwilliams)<a name="note_2"></a>|@markrwilliams commented|
|-|-|

Replying to [adiroiban](#note_1):
> and as part of this ticket we should also put this new API in use by fixing the twistd non-ascii stderr handling...
> 
> I think that this ticket should focus on fixing that bug and create the compat helper as part of the fix

Sounds good to me.  I will merge [#8155](https://github.com/twisted/twisted/issues/8155) and then update the PR with that fix.  Does that sound OK?

**Comment by twisted-trac** [comment]
|[<img alt="markrwilliams's avatar" src="https://avatars.githubusercontent.com/u/704247?s=50" width="50" height="50">](https://github.com/markrwilliams)<a name="note_3"></a>|@markrwilliams commented|
|-|-|

PR:
https://github.com/twisted/twisted/pull/470

**Comment by twisted-trac** [comment]
|[<img alt="markrwilliams's avatar" src="https://avatars.githubusercontent.com/u/704247?s=50" width="50" height="50">](https://github.com/markrwilliams)<a name="note_4"></a>|@markrwilliams commented|
|-|-|

I fixed the issue from [#8155](https://github.com/twisted/twisted/issues/8155) and also fixed another issue, where non-bytes were being written to a pipe.

**Comment by twisted-trac** [comment]
|[<img alt="markrwilliams's avatar" src="https://avatars.githubusercontent.com/u/704247?s=50" width="50" height="50">](https://github.com/markrwilliams)<a name="note_6"></a>|@markrwilliams commented|
|-|-|

I think that using `networkString` to encode the error reported through `postApplication` means it's not possible to receive non-ASCII data over the status pipe.  That would mean it's not necessary to present the repr of the bytes received.

Would it be better to encode the data using the system's default encoding before writing to the status pipe, and then always show the bytes repr on the read side?

**Comment by twisted-trac** [comment]
|[<img alt="markrwilliams's avatar" src="https://avatars.githubusercontent.com/u/704247?s=50" width="50" height="50">](https://github.com/markrwilliams)<a name="note_7"></a>|@markrwilliams commented|
|-|-|

Or maybe it's better to use `surrogateescape` on Python 3?

**Comment by twisted-trac** [comment]
|[<img alt="markrwilliams's avatar" src="https://avatars.githubusercontent.com/u/704247?s=50" width="50" height="50">](https://github.com/markrwilliams)<a name="note_8"></a>|@markrwilliams commented|
|-|-|

I've moved the PR to https://github.com/twisted/twisted/pull/476

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_9"></a>|@adiroiban set owner to @markrwilliams|
|-|-|

Thanks for the work on this.

Changes look good.

I think that the ticket and description of this ticket should be updated to talk about the problems that we get with non-ascii communication between parent and child process during a fork.

the fact that we introduced a private helper in compat should be of secondary importance :)

Please see my inline minor comments for the PR (and the comments from the other contributors) and merge.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_12"></a>|@adiroiban commented|
|-|-|

Great update.

` Add helper to create bytes reprs that always begin with b.` ... this might be confusing, since it does not tell the name of the helper... and the helper is in fact private, so end users should not use it.

Just leave it out.

Please check my comment and merge.

Many thanks!

**Comment by twisted-trac** [comment]
|[<img alt="markrwilliams's avatar" src="https://avatars.githubusercontent.com/u/704247?s=50" width="50" height="50">](https://github.com/markrwilliams)<a name="note_13"></a>|@markrwilliams set status to `closed`|
|-|-|

In changeset e3140e79ae8b4575a40e1bfbf0dfcc06103b0a72
```
#!CommitTicketReference repository="" revision="e3140e79ae8b4575a40e1bfbf0dfcc06103b0a72"
Merge 8731-bytes-repr-with-b: Correct error reporting in _twistd_unix.py

Author: markrwilliams
Reviewer: adiroiban, rodrigc
Fixes: #8731

Errors during daemonization are now correctly reported on all Python
versions.
```
